### PR TITLE
Java Finalize Class is a removal candidate in JDK 18 #22260

### PR DESCRIPTION
**Pull request by nroduit**
This PR proposes an important change to better manage the release of native memory related to JAVA classes.

See more details at https://github.com/opencv/opencv/issues/22260

Objectives:
- Allows replacing properly finalize(). Enable compilation from JDK 18.
- Allow a better memory release during intensive use of images. In some cases can lead to a crash of the JVM.

Changes:
- Separate Java Classes and sources to make possible the multi-release package: e.g. opencv-470.jar and opencv-470-sources.jar
- The output binary package is a [multi-release jar](https://openjdk.org/jeps/238) that is Java 8 compatible but if the JVM is higher than Java 8 then it will use the Java 9 classes that are set in "META-INF/versions/9/org/opencv". So it allows the distribution of a Java 8+ compatible package where only JVM 9+ will use Java 9 classes. This means that using JDK 8 in runtime does not bring any benefits (same as the previous situation).

As the changes can have impacts across all modules, it is not excluded that there are side effects. Especially Android has not been tested.

This patch has already been deployed in production. No problems were reported after several weeks of intensive reading or writing of images. 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x ] I agree to contribute to the project under Apache 2 License.
- [ x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by nroduit** [comment]
Runner prerequisites: JDK 11 and a recent version of ANT

**Comment by asmorkalov** [comment]
@nroduit Thanks a lot for the contribution! Could you reduce code duplication for Mat class. Options I see:
- Use inheritance. No need to copy all Mat members twice.
- Stay with single class and presume finalize method, but check Java version in runtime: new java set cleaner and does nothing in finalize(), old - vice versa.
- Some hacks with preprocessor, e.g. cpp, but it looks ugly and leads to maintenance issues.

**Comment by nroduit** [comment]
I thought that the Java API distributed by OpenCV was compatible with Java 8 but I see that in the packages of release 4.7.0, the minimal version is Java 11.

If the strategy is a minimum version of Java 11, then the modifications should be easier. The duplication of the Mat class is not necessary anymore.
Otherwise, I don't think it's possible not to duplicate the code of the Mat class because you have to delete "finalize()". This is not possible with inheritance and moreover, this function comes from java.lang.Object which cannot be compiled with a JDK 18+ (with default settings).

Thanks for specifying the strategy.

**Comment by asmorkalov** [comment]
Android is yet another Java API consumer. Please take it into account.

**Comment by asmorkalov** [comment]
@nroduit  Thanks a lot for the contribution! I dig a bit in details and have alternative solution in mind.

Motivation: Reduce code duplication and do not copy Core.Mat implementation several times. Also handle as less copies of classes as possible in OpenCV SDK.

Proposal:
- Extract cleaner from Core.Mat as separate entity in Core and wrap it with own class.
- Replace `org.opencv.core.Mat.cleaner.register(this, () -> delete(addr));` with:
```
String version = System.getProperty("java.version");
if (!version.startsWith("1.")) {
   org.opencv.core.Cleaner.cleaner.register(this, () -> delete(addr));
}
```
- Presume `finalize` method with version check:
```
String version = System.getProperty("java.version");
if (version.startsWith("1.")) {
   delete(nativeObj);
}
```

Pros:
- No need to maintain branches in generator. Only `Cleaner` is Java version specific.
- No need to maintain two versions of Core.Mat.
- Minimal release infrastructure changes. Finalize method may be removed any time easily.

Cons:
- Fake `finalize ` may be trigger for linters.
- Version check may take time. It may be done once and extracted as Cleaner field.

**Comment by nroduit** [comment]
I understand the intention of not wanting to duplicate the code, but the problem resides elsewhere.

In your proposal, there will be a compilation error with JDK 18 and higher (see https://github.com/opencv/opencv/issues/22260). As the [finalize method can no longer be declared](https://www.infoworld.com/article/3649089/how-to-handle-java-errors-and-cleanup-without-finalize.html), we can't envisage the 2 solutions cohabiting in the same code. That's why I can't see any other solution than to manage Java 8 and Java 9+ classes separately.

However, if the minimal compilation version in OpenCV becomes Java 11, then there's no need to duplicate the code and implementation can be carried out directly using the Cleaners method. As my understanding is limited to a few OpenCV modules, I'm not fully aware of all the consequences, especially for the Android part.

In short, I can contribute to making the code compatible with recent versions of Java and making memory release more efficient. However, I can't decide on the technological life cycles in OpenCV and I need help to finalize this PR to ensure that Java works properly with Android (which I've never used).

**Comment by asmorkalov** [comment]
OpenCV supports Java for Android too. Minimal API level for cleaners support is 33 and it's very high requirement.

**Comment by AlexanderSchuetz97** [comment]
I would recommend implementing what you have implemented with a reference queue.
This should work with even ancient java versions and there are no plans to deprecate.
"Cleaner" is not really needed to solve this problem, nor does it provide any advantage.

I would propose the following:
Create a new class called MatRef that holds the address long and extends Phantom or Weak Reference.
MatRef becomes a member of Mat instead of address.
Whenever Mat tries to access the address it will use this.ref.address instead of this.address.

MatRef is enqueued on a global static ReferenceQueue and has the Mat object as a Referent. 

Then create a global thread thats started when opencv is loaded calling remove constantly on the global reference queue. This thread will fetch the MatRef instances from the Reference Queue. (This will require a cast)

Whenever a MatRef is polled from the Queue it means that the Mat Object that used to hold the MatRef is gone. All thats left to do is call the delete method with the long.

If you want to make it "fancy" then you make MatRef.adress private and give us a Mat.free() method which will instantly perform the cleanup and add a MatRef.addess() method that will throw NullPointerException perhaps if the Mat has already been freed. 

All of this will probably work on any java version starting with version 6 and fix most of the memory management issues that exist with opencv. I personally hate having to constantly call System.gc() to prefent my applications from running out of memory. Doing something like this would allow me to call free() explicitly and still not break existing code that does not call free().

This also does not consume more Memory than using the generic finalize() method, as the JVM will create a companion Reference object for every java Object with a finalize() method in existance. (This is how the jvm calls the finalize() method)
All of this happens in the background and can only be observed in a HeapDump.

I am not a Android dev and don't know how this would work in Android but I highly doubt that android does not implement ReferenceQueue.

**Comment by nroduit** [comment]
@AlexanderSchuetz97 your explanation corresponds exactly to the implementation of java.lang.ref.Cleaner available since Java 9.

I therefore implemented CustomCleaner, which is similar to java.lang.ref.Cleaner, and removed all finalize() methods.
As described in [CustomCleaner](https://github.com/nroduit/opencv-1/blob/java-finalize/modules/core/misc/java/src/java/CustomCleaner.java#L8-L18), when Java 8 support will be discontinued, CustomCleaner can be simply replaced with java.lang.ref.Cleaner.

I've simplified the modifications in the python class generation script to reduce the risk of side-effects in all other generated classes.

> If you want to make it "fancy" then you make MatRef.adress private and give us a Mat.free() method which will instantly perform the cleanup and add a MatRef.addess() method that will throw NullPointerException perhaps if the Mat has already been freed.

To free memory before the object is dereferenced by the garbage collector, use the Mat.release() method, which frees the content but keeps the pointer. 
It is important not to call Cleaner.clean() several times because it calls n_delete() and this usually crashes the JVM. 

I've also set a minimal language level in the Ant build by setting Java 8 explicitly because the added code requires Java 8 and because, given the end-of-life cycle of [Java](https://endoflife.date/java) and [Android](https://endoflife.date/android), it seems reasonable to me to no longer support lower versions.

**Comment by nroduit** [comment]
Remove Java explicit version in Ant which seems to be passed by command with OPENCV_JAVA_SOURCE_VERSION and OPENCV_JAVA_TARGET_VERSION

Note also that I've separated class and source into 2 separate jars. The source jar contains the "-sources" qualifier, which is the convention used with Maven repositories.

**Comment by nroduit** [comment]
@asmorkalov, in Android test, the minimal [API version is 21](https://github.com/opencv/opencv/blob/6694d87a23886182900cdb49876d98989cc6e16d/platforms/android/ndk-18-api-level-21.config.py), it must be API 26 to support Java 8 lambda. API 26 is Android 8 which has no security update for more than 2 years.

**Comment by nroduit** [comment]
Any feedback on this error? @asmorkalov

**Comment by nroduit** [comment]
Any news on this problem? Shouldn't API compatibility evolve with version 5?

API 21 (8.0 Oreo) has been [end-of-life for over 5 years ](https://endoflife.date/android)and no longer receives security patches.

**Comment by AlexanderSchuetz97** [comment]
Just like nroduit I also need this.
Future Java Versions will soon no longer call finalize at all.

File: modules/core/misc/java/src/java/CustomCleaner.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+        private final Runnable action;
+       PhantomCleanable prev = this;
+       PhantomCleanable next = this;
+
+        private final PhantomCleanable list;
+
+        public PhantomCleanable(Object referent, Runnable action) {
+            super(Objects.requireNonNull(referent), queue);
+            this.list = phantomCleanableList;
```
Comment (opencv-alalek) [inline modules/core/misc/java/src/java/CustomCleaner.java:93]: Why do we need additional "list" if there is already "queue"?

Any link on "reference implementation code"?

**Comment by nroduit** [comment]
@asmorkalov Can you review

File: modules/core/misc/java/src/java/CustomCleaner.java (hunk 1 above)
Comment (AlexanderSchuetz97) [reply modules/core/misc/java/src/java/CustomCleaner.java:93]: I am not the Author of this code, but I feel qualified to comment anyways.

This is the correct way to implement this.

You need to ensure that the cleaner itself outlives the resource you want to clean.
The "ReferenceQueue" itself does not do that. Therefore you need a different datastructure like a list or hashmap.
It needs to be a double linked list or hashmap because you have no idea in what order the elements are to be removed. So you need to support fast removal of elements in the center. Making the cleaners themselves (they know when to be removed) the node in a  list as done here is the canonical way to do it.

In the link you posted, which is a short example, the caller ensures that the cleaner reference remains on the stack and thereforce not garbage collected. This approch is fine if you know how many elements you need to clean (in the case of the example that number is 1.) However this is not possible for opencv's case you have no knowledge on how many mat's are create and you also do not want to force ever user of opencv to change their code to also keep the cleaner alive for longer than the mat.

File: modules/core/misc/java/src/java/CustomCleaner.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    public Cleanable register(Object obj, Runnable action) {
+        return new PhantomCleanable(Objects.requireNonNull(obj), Objects.requireNonNull(action));
+    }
+
+    private void start() {
+        new PhantomCleanable(this, null);
+        Thread thread = new CleanerThread(() -> {
+            while (!phantomCleanableList.isListEmpty()) {
+                try {
+                    Cleanable ref = (Cleanable) queue.remove(60 * 1000L);
```
[the remark this answers is not in the document]
Comment (AlexanderSchuetz97) [reply modules/core/misc/java/src/java/CustomCleaner.java:45]: Nroduit uses a long polling pattern here. Basically as explained above phantomCleanableList.isListEmpty() is only true when no reference to CustomCleaner exists. And if that is the case then the thread should stop because without a reference to the CustomerCleaner object enquening further cleaners is impossible.

My thoughts on this implementation:
Its not how I would have implemented this, but its a valid way of doing it, possibly even the better way.
Nroduit has implemented this entire thing in a way that multiple cleaners and therefore multiple cleaner threads can be created. Openjdk/Oracle? recommends (somewhere in the jdk 9 'Cleaner' javadoc, which is what we are trying to re-create here) to only use exactly 1 cleaner per library. I would have not bothered with this complexity of implementing "stopping" of the cleaner and made it never stop once started. Then you wouldnt need a timeout. The only scenario I can think of where stopping would theoretically be required is if you wanted to support class unloading of opencv. I highly doubt that this would be a good idea as I dont think you guys properly implemented JNI_OnUnload on the native side to not leak memory. I think 95% of people dont use class unloading in their applications anyways, so its not really a problem.

**Comment by asmorkalov** [comment]
Rebased on top of current 4.x to trigger CI and refresh bindings code.

File: modules/core/misc/java/src/java/CustomCleaner.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    public Cleanable register(Object obj, Runnable action) {
+        return new PhantomCleanable(Objects.requireNonNull(obj), Objects.requireNonNull(action));
+    }
+
+    private void start() {
+        new PhantomCleanable(this, null);
+        Thread thread = new CleanerThread(() -> {
+            while (!phantomCleanableList.isListEmpty()) {
```
Comment (vlsi) [inline modules/core/misc/java/src/java/CustomCleaner.java:43]: This is likely invalid: once the thread terminates it no longer can process further references.

If a user creates opencv object, releases it, then waits for 120 seconds, then allocates yet another opencv object the second object will never be cleared.
The cleaner thread should restart in that case.

File: modules/core/misc/java/src/java/CustomCleaner.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        return customCleaner;
+    }
+
+    public Cleanable register(Object obj, Runnable action) {
+        return new PhantomCleanable(Objects.requireNonNull(obj), Objects.requireNonNull(action));
+    }
+
+    private void start() {
+        new PhantomCleanable(this, null);
+        Thread thread = new CleanerThread(() -> {
```
Comment (vlsi) [inline modules/core/misc/java/src/java/CustomCleaner.java:42]: `new Thread` captures `Thread.inheritedAccessControlContext` so the code will likely run into memory leak like in https://stackoverflow.com/questions/78163385/does-postgres-jdbc-lazycleaner-cause-a-classloader-memory-leak/79835942

**Comment by vlsi** [comment]
Have you considered using a tested dependency rather than reinventing the cleaner? (e.g. see https://github.com/juanplopes/lazycleaner ).

---

I would suggest updating the API so the user could call something like `Mat.close()` or `Mat.free()` and make it free the native resources and exclude the object from the cleaner queue.
